### PR TITLE
New jenkins plugin fails with confusing message when used with older IQ Servers CLM-10117 CLM-10103 

### PR DESCRIPTION
**Pull request by ja98200wilcox**
Provide an error message when the IQ server and Jenkins plugin are not compatible.

https://jenkins.zion.aws.s/job/integrations/job/jenkins/job/sonatype-nexus-platform-plugin-feature/job/CLM-10103_Need_better_msg_for_unsupported_IQ/

https://issues.sonatype.org/browse/CLM-10117

### CODE REVIEW

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 1]
```diff
@@ -91,13 +92,31 @@ class IqPolicyEvaluatorUtil
...
       return null
     }
   }
 
+  private static boolean isHttp404Error(final Exception throwable) {
+    boolean result = false
+    int httpResponseExceptionIndex = ExceptionUtils.indexOfType(throwable, HttpResponseException)
+    if (httpResponseExceptionIndex >= 0) {
+      Throwable[] throwables = ExceptionUtils.getThrowables(throwable)
+      HttpResponseException httpResponseException = (HttpResponseException) throwables[httpResponseExceptionIndex]
```
Comment (tneer) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: I'm being lazy and not looking this up myself on the hope that you know from your research. Do we know for sure that this is going to be a HttpResponseException or do we need to do an instanceof check before casting it?

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 2]
```diff
@@ -91,13 +92,31 @@ class IqPolicyEvaluatorUtil
...
+      if (httpResponseException.statusCode == 404) {
```
Comment (tneer) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: Minor.  Personal pet peeve.  I think you could just

```java
return (httpResponseException.statusCode == 404)
```
Maybe get rid if the result variable altogether and just return false if we fell through to there?

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties [hunk 3]
```diff
@@ -30,6 +30,7 @@ IqPolicyEvaluation.FailOnNetwork=Fail build when unable to communicate with IQ S
 IqPolicyEvaluation.JobSpecificCredentials=Use job specific credentials
 
 IqPolicyEvaluation.NoIqServersConfigured=No IQ Server configured.
+IqPolicyEvaluation.ResourceNotFound=Possible version mismatch.  Please check if the IQ Server is compatible with this plugin.
```
Comment (tneer) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: Minor:  Maybe reword to:

Possible version mismatch.  Please check if the IQ Server is compatible with this plugin version.

File: src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy [hunk 4]
```diff
@@ -394,4 +395,30 @@ class IqPolicyEvaluatorTest
...
+      buildStep.perform(run, workspace, launcher, listener)
+
+    then:
+      noExceptionThrown()
+      1 * run.setResult(Result.FAILURE)
+      1 * logger.println({ Messages._IqPolicyEvaluation_ResourceNotFound().toString() })
+
+    where:
+      exception                                         | failBuildOnNetworkError || expectedException | expectedMessage
+      new IqClientException('BOOM!!',
```
Comment (tneer) [inline src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy]: :-)

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 5]
```diff
@@ -91,13 +92,31 @@ class IqPolicyEvaluatorUtil
     if (!isNetworkError || failBuildOnNetworkError) {
       throw e
     }
+    else if (isHttp404Error(e)) {
```
Comment (fjmilens3) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: I think it would be better if somehow we could implement this _only_ for the new API operation; I don't know if we want _every_ 404 to produce a "possible version mismatch" message since that seems incredibly broad in the general case, at least if it's not too difficult to make it more scoped.

Could we just add this particular check as part of a try-catch for the new operation, delegating to what we already had if it's not a 404?

https://github.com/sonatype/nexus-platform-plugin/blob/39b293a3c0e390091a910feeb6ea63374b213ace/src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy#L52

Another idea, perhaps a bit hackier, would be to extend the `handleNetworkException` method with a `boolean` that would indicate whether or not the message should mention a possible version mismatch; you could declare the boolean at the start of the method as false, set it to true when you make the `verifyOrCreateApplication` call, and then set it back to false once that particular operation had concluded.

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties (hunk 3 above)
Comment (fjmilens3) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: If we're going to use this for all 404s, I might extend the message a bit, such as:

`Required resource not found, possible version mismatch. Please check if the IQ Server is reachable and compatible with this plugin.`

At least that way we cover both options, I'd hate for something basic like an unreachable server because of some internal network weirdness to have someone spend a day thinking they have the wrong version. I know they should check such things, but if you're a techie in a pressure-cooker environment, I think it's at least possible they'll take the error message for more than they should, and it could lead them astray.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 1 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > Do we know for sure that this is going to be a HttpResponseException

Yes, I know for sure.   The call ```int httpResponseExceptionIndex = ExceptionUtils.indexOfType(throwable, HttpResponseException)``` is why i know for sure.  I first check is it a ```HttpResponseException``` by looking if it is a member of the throwables.  If yes i then grab it and cast it.

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties (hunk 3 above)
Comment (vladt) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: Since this is implemented for a very specific case, I would like to see it limited to that case, which would allow us to have a very specific message with the actual IQ server version that is required.

We have another Jira for genericmismatches between server and plugin versions .

Comment (ja98200wilcox) [reply src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: > Maybe reword to

changed in 02dd0fa

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 5 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: >  It should be only for the new call to verifyOrCreateApplication.

Changed

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 2 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > Maybe get rid if the result variable altogether and just return false if we fell through to there?

Changed e299be4

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties (hunk 3 above)
Comment (ja98200wilcox) [reply src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: > Since this is implemented for a very specific case, I would like to see it limited to that case, which would allow us to have a very specific message with the actual IQ server version that is required.

Changed 31c1544

Comment (ja98200wilcox) [reply src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: > If we're going to use this for all 404s, I might extend the message a bit, such as:

Changed message to be specific e299be4

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 6]
```diff
@@ -98,6 +111,16 @@ class IqPolicyEvaluatorUtil
     }
   }
 
+  private static boolean isHttp404Error(final Exception throwable) {
+    int httpResponseExceptionIndex = ExceptionUtils.indexOfType(throwable, HttpResponseException)
+    if (httpResponseExceptionIndex >= 0) {
+      Throwable[] throwables = ExceptionUtils.getThrowables(throwable)
+      HttpResponseException httpResponseException = (HttpResponseException) throwables[httpResponseExceptionIndex]
+      return httpResponseException.statusCode == 404
```
Comment (DennisWarren) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: Instead of hard coding we could use HttpStatus.NOT_FOUND_404

File: src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy [hunk 7]
```diff
@@ -394,4 +395,30 @@ class IqPolicyEvaluatorTest
       1 * log.println('\nThe detailed report can be viewed online at http://server/report\n' +
           'Summary of policy violations: 0 critical, 0 severe, 0 moderate')
   }
+
+  def 'Resource not found exception handling'() {
```
Comment (fjmilens3) [inline src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy]: I would change the name of the test to reflect the more specific behavior of the code.

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties [hunk 8]
```diff
@@ -30,6 +30,7 @@ IqPolicyEvaluation.FailOnNetwork=Fail build when unable to communicate with IQ S
 IqPolicyEvaluation.JobSpecificCredentials=Use job specific credentials
 
 IqPolicyEvaluation.NoIqServersConfigured=No IQ Server configured.
+IqPolicyEvaluation.AutoAppMissing=Missing Auto App Creation.  Possible version mismatch. Please check if the IQ Server is compatible with this plugin version.  Supported IQ server versions are 1.45.0 or newer.
```
Comment (fjmilens3) [inline src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: Minor: Extra space between the first period and `Possible`.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 6 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > Instead of hard coding we could use HttpStatus.NOT_FOUND_404

Changed c37de51

File: src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy (hunk 7 above)
Comment (ja98200wilcox) [reply src/test/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorTest.groovy]: > I would change the name of the test to reflect the more specific behavior of the code.

Changed 2be7c0a

File: src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties (hunk 8 above)
Comment (ja98200wilcox) [reply src/main/resources/org/sonatype/nexus/ci/iq/Messages.properties]: >Minor: Extra space between the first period and Possible.
Minor: Extra space between the period after version and Supported.



Changed 2580539

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 9]
```diff
@@ -49,8 +51,20 @@ class IqPolicyEvaluatorUtil
           new IqClientFactoryConfiguration(credentialsId: iqPolicyEvaluator.jobCredentialsId, context: run.parent,
               log: loggerBridge))
 
-      def verified = iqClient.verifyOrCreateApplication(applicationId)
-      checkArgument(verified, 'The application ID ' + applicationId + ' is invalid.')
+      try {
+        def verified = iqClient.verifyOrCreateApplication(applicationId)
+        checkArgument(verified, 'The application ID ' + applicationId + ' is invalid.')
+      }
+      catch (IqClientException e) {
```
Comment (vladt) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: I would move the error handling to DefaultIqClient.verifyOrCreateApplication. I think it will be simpler there.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 10]
```diff
@@ -49,8 +51,20 @@ class IqPolicyEvaluatorUtil
...
+        if (isHttp404Error(e)) {
+          listener.logger.println Messages._IqPolicyEvaluation_AutoAppMissing()
+          run.result = Result.FAILURE
+          return null
```
Comment (vladt) [inline src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: What happens after you return null here? Isn't that still an exception that must be thrown?

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 9 above)
Comment (DennisWarren) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: >I would move the error handling to DefaultIqClient.verifyOrCreateApplication. I think it will be simpler there.

I think that might be a good idea as well.  We have these api calls that basically remove the need for http response handling, but now we have to go into the response and pull out the status code.  It would be nice to push this higher up so that we can make plain `iqClient.verifyOrCreateApplication` calls

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 10 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > What happens after you return null here? Isn't that still an exception that must be thrown?

No exception is thrown.  I inform Jenkins it failed by setting the ``run.result = Result.FAILURE```.   If I did throw an exception it will be displayed instead of my message.   If I let the exception be thrown it would change the output to be the stacktrace.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 9 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > I would move the error handling to DefaultIqClient.verifyOrCreateApplication. I think it will be simpler there.

I do believe the implementation for verifyOrCreateApplication is located outside of the jenkins plugin.  It is located in another project.  I think its the java-api project.  Also, all the other calls throw an IqClientException and the plugin looks into the throwables to see what the underlying exception is.  I am not saying that is a good thing.  I just feel changing the way exceptions are handled which are generated from the java-api project is outside of the scope of this PR.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy [hunk 11]
```diff
@@ -49,8 +50,20 @@ class IqPolicyEvaluatorUtil
           new IqClientFactoryConfiguration(credentialsId: iqPolicyEvaluator.jobCredentialsId, context: run.parent,
               log: loggerBridge))
 
-      def verified = iqClient.verifyOrCreateApplication(applicationId)
-      checkArgument(verified, 'The application ID ' + applicationId + ' is invalid.')
+      try {
```
[the remark this answers is not in the document]
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > I do wonder if it would be cleaner at this point to extract all this to its own method

I could move all the logic into another method.  Return void and the next call after calling this method could check the run.status.  if it is bad, do a return null.  I see that being done in other parts of the code.

Example: 

```
      verifyOrCreateApplication(iqClient, listener, run, applicationId)
      if (run.getResult() == Result.FAILURE) {
        return null
      }
```
The method:

```
private static void verifyOrCreateApplication(final InternalIqClient iqClient, final TaskListener listener, final Run run, final String applicationId) {
    try {
      def verified = iqClient.verifyOrCreateApplication(applicationId)
      checkArgument(verified, 'The application ID ' + applicationId + ' is invalid.')
    }
    catch (IqClientException e) {
      if (isHttp404Error(e)) {
        listener.logger.println Messages._IqPolicyEvaluation_AutoAppMissing()
        run.result = Result.FAILURE
      }
      else {
        throw e
      }
    }
  }
  ```

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 9 above)
Comment (fjmilens3) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > I am not saying that is a good thing. I just feel changing the way exceptions are handled which are generated from the java-api project is outside of the scope of this PR.

Personally, I'll take that as a valid answer. If we have to modify that much to change this message I don't think that's something that was anticipated when we sized this, and even if it's a trivial change, it does add a lot more time in terms of completing the story.

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 10 above)
Comment (vladt) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > No exception is thrown. I inform Jenkins it failed by setting the run.result = Result.FAILURE. If I did throw an exception it will be displayed instead of my message. If I let the exception be thrown it would change the output to be the stacktrace.

Why is that undesired? Why not throw an exception with the new message and the cause set to the caught exception?

File: src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy (hunk 9 above)
Comment (ja98200wilcox) [reply src/main/java/org/sonatype/nexus/ci/iq/IqPolicyEvaluatorUtil.groovy]: > I was thinking that we could move the exception handling to DefaultIqClient.verifyOrCreateApplication, which will throw now an IqClientException with the new message. 

That sounds good to me.  I misunderstood your suggestion before.  But now I understand what you are saying and I like it.  I would add this message to the DefaultIqclient and I wouldn't need to change anything in the Jenkins project besides updating the POM.  Jenkins already has the ability to show the exception.
